### PR TITLE
Optimize PREWHERE

### PR DESCRIPTION
**Pull request by amosbird**
So that it's never worse than WHERE.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):

Now prewhere should be at least as efficient as where. As a result, any viable predicates can be moved into prewhere without fancy heuristic

```
before:

    "runs": [
        {
            "bytes_per_second": 800334988.343924,
            "memory_usage": 0,
            "min_time": 1.290000,
            "quantiles": {
                "0.1": 1.299352,
                "0.2": 1.301392,
                "0.3": 1.303418,
                "0.4": 1.304222,
                "0.5": 1.305335,
                "0.6": 1.311000,
                "0.7": 1.324643,
                "0.8": 1.328232,
                "0.9": 1.338506,
                "0.95": 1.344474,
                "0.99": 1.349096,
                "0.999": 1.350136,
                "0.9999": 1.350240
            },
            "queries_per_second": 0.722716,
            "query": "SELECT Title, URL FROM hits_10m_single PREWHERE WatchID % 2 = 1 WHERE UserID = 10000 FORMAT Null",
            "query_index": 0,
            "rows_per_second": 7568227.234566,
            "total_time": 17.987692
        }
    ],


after:

    "runs": [
        {
            "bytes_per_second": 1963621034.843018,
            "memory_usage": 0,
            "min_time": 1.030000,
            "quantiles": {
                "0.1": 1.037758,
                "0.2": 1.043031,
                "0.3": 1.045489,
                "0.4": 1.053987,
                "0.5": 1.067302,
                "0.6": 1.077478,
                "0.7": 1.080989,
                "0.8": 1.088930,
                "0.9": 1.118968,
                "0.95": 1.123727,
                "0.99": 1.127534,
                "0.999": 1.128391,
                "0.9999": 1.128477
            },
            "queries_per_second": 0.892347,
            "query": "SELECT Title, URL FROM hits_10m_single PREWHERE WatchID % 2 = 1 WHERE UserID = 10000 FORMAT Null",
            "query_index": 0,
            "rows_per_second": 9246116.658452,
            "total_time": 11.206397
        }
    ],
```

### CODE REVIEW

**Comment by zhang2014** [comment]
How does this work when the `where expression` filter  large amounts of data ? 
in our scenario, we always put `where expression` that are not included in the `primary key` into `prewhere expression`.

**Comment by amosbird** [comment]
PREWHERE in ClickHouse can serve two optimization purposes.

1. Reduce disk read IO

This is the main purpose of PREWHERE. When the prewhere predicate has low selectivity (f.g. 10%), and non prewhere columns to be scanned is large, PREWHERE reduces a large amount of disk read IO, and saves some efforts to do filtering. But, if the prewhere predicate has high selectivity (f.g. 90%), and there exists another WHERE predicate, we might waste too much time pre-filtering a large number of columns while only decrease a small amount of disk read time. As a result, we have heristics to not push everything from WHERE into PREWHERE.

However, in order to reduce disk IO, we don't need the filter action, only the filter vector needs to be calculated (and can be reused afterwards). This is what this pr does. It calculates the filter vector, scanning non-prewhere columns based on that, and rewritting `WHERE` as `AND(prewhere_filter, WHERE)`. Then we achieve this goal without potential regressions.

2. Lower CPU usage

ClickHouse doesn't support AND short-circuit evaluation, https://github.com/ClickHouse/ClickHouse/issues/1562 . It's a hard problem for vectorized engines. Therefore, predicate order doesn't affect performance. However, after we introduce PREWHERE, it adds an additional stage to predicate evaluation, which could be utilized for predicate order optimization. For example, suppose we have two predicate functions, `foo(a)` and `bar(b)`, with runtime complexity Θ(1) and Θ(1000) per block respectively, and suppose filtering a block of (a, b) takes O(10). If we place `foo(a)` into PREWHERE, even when `foo`'s selectivity is 90%, we still gain O(90). That is, double filtering is effcient as long as the latter predicates are expensive.

The tricky thing in this situation is that we have two metrics which are hard to estimate, a) the cost of filtering a block, and b) the cost of evaluating predicates. And it can be data related such that strategies are different per block. Thus further researches are required. This pr doesn't filter anything in PREWHERE, thus should have no regression compared to WHERE.

**Comment by amosbird** [comment]
https://github.com/yandex/ClickHouse/blob/master/dbms/tests/queries/1_stateful/00091_prewhere_two_conditions.sql#L6   catches another tradeoff. If we put predicates with heavy columns into PREWHERE, we might increase IO

**Comment by alexey-milovidov** [comment]
Do I understand correctly, that:
- we avoid filtering data twice if the filter in PREWHERE does not filter much?

> As a result, any viable predicates can be moved into prewhere without fancy heuristic

But we already doing it (almost: we exclude conditions on the first primary key column).
And the choice of conditions to move still depend on the relative difficulty of reading columns and the guess on selectivity.

**Comment by alexey-milovidov** [comment]
Now we can go further and implement:

1. Do "prewhere" style filtering in MergeTree tables without modifying query AST (without actually involving PREWHERE).
2. Implement multi-step "prewhere" filtering when we read columns and evaluate conditions one by one, deciding whether to do filtering on each step.

Multistep prewhere is our long appreciated goal!

**Comment by alexey-milovidov** [comment]
I've merged this PR, but there are some drawbacks:

1. There is no performance tests added.
2. The threshold 0.6 is not controlled by setting and there is no comments about how it was chosen.
3. Some style issues like `num_rows = total_rows_per_granule = new_filter->size();`

Could you please address these is follow-up PR?

**Comment by alexey-milovidov** [comment]
If the issue #8474 is fixed, could you please also add a test?

**Comment by amosbird** [comment]
> But we already doing it (almost: we exclude conditions on the first primary key column).

This pr started to optimize away the heuristic of MergeTreeWhereOptimizer, but fails due to consequent findings during optimizing prewhere. Multistep prewhere might be only one dimension of the solution space to the problem. Which columns to read first depends at least on 

1. the actual amount of IO to read it
2. the actual amount of cpu resource to compute filters related to it
3. the benefits to other subsequent filters due to excessive amount of IO spent on it.

Also we might need block with unaligned columns (column with different row numbers) to avoid multi-pass filtering. Also a dedicated materialized representation of const-column will benefit.

So back to this pr. It does one optimization: to avoid double filtering when PREWHERE filters cannot reduce significant amount of IO in current block, that is, we defer the filtering to WHERE. The process is per block, hence it still helps reduce IO if possible.

It also fixes two bugs in prewhere, one is later found in https://github.com/ClickHouse/ClickHouse/issues/8474 , another is more serious, causing segfault due to mis deduced type of predicate column.
```
CREATE TABLE default.ml (`a` LowCardinality(String), `x` Nullable(Int32)) ENGINE = MergeTree ORDER BY tuple()

insert into ml values ('', 2);

select a, y  from ml prewhere (x = 2) as y
```

It also changes the behavior of `max_bytes_to_read`. Before this pr it counts the byte after prewhere filteration, which however, doesn't reflect the real IO. From the setting's comment : `Limit on read bytes ...` . It should come before prewhere filteration. This is catched by test  00091_prewhere_two_conditions.sql


> There is no performance tests added.

tests are added in different pr https://github.com/ClickHouse/ClickHouse/pull/7768

> The threshold 0.6 is not controlled by setting and there is no comments about how it was chosen.

same as https://github.com/amosbird/ClickHouse/blob/c%2B%2B20/dbms/src/Storages/MergeTree/MergeTreeRangeReader.cpp#L278  and https://github.com/amosbird/ClickHouse/blob/c%2B%2B20/dbms/src/Storages/MergeTree/MergeTreeWhereOptimizer.cpp#L177

This not a perfect solution as I expected due to dilemmas described in above comments, so I don't have a support for that value, it's choosed as the minimal score a student need to received to pass the final exam... 60 / 100.

> If the issue #8474 is fixed, could you please also add a test?

will do

File: dbms/src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -412,9 +412,19 @@ InterpreterSelectQuery::InterpreterSelectQuery(
             query.setExpression(ASTSelectQuery::Expression::WHERE, std::make_shared<ASTLiteral>(0u));
         need_analyze_again = true;
     }
+    if (query.prewhere() && query.where())
+    {
+        /// Filter block in WHERE instead to get better performance
+        query.setExpression(ASTSelectQuery::Expression::WHERE, makeASTFunction("and", query.prewhere()->clone(), query.where()->clone()));
+        need_analyze_again = true;
```
Comment (azat) [inline dbms/src/Interpreters/InterpreterSelectQuery.cpp:419]: Looks like this optimization removed by #42126 (cc @davenger ) and there is no need in this code anymore, let's try - #54535

Comment (azat) [reply dbms/src/Interpreters/InterpreterSelectQuery.cpp:419]: Hm, actually this is not true, it does not apply PREWHERE always

**Pull request merged**